src/ophys_etl/modules/segmentation/merge/roi_merging.py

```diff
@@ -0,0 +1,378 @@
+        timeseries that is a weighted average of the pixels in roi0
+        using the brightest filter_fraction of time steps in the
+        characteristic timeseries. Use these correlations to construct
+        a Gaussian distribution.
+
+        2b) Correlate all of the pixels in roi1 against the characteristic
+        timeseries for roi0 using the same timesteps.
+
+        2c) Calculate the z-score of the correlations from (2b) using
+        the Gaussian distribution from (2a).
+
+        2d) Reverse roi0 and roi1 and repeat steps (2a-c). Evaluate the
+        merger based on the highest resulting median z-score. If that
+        z-score is greater than or equal to -1*corr_acceptance,
+        the merger is valid.
+
+    3) Rank the potential valid mergers based on the median z-sccore
+    from step (2d). Move down the list of ranked mergers, merging ROIs.
+    Once an ROI participates in the merger, it is removed from
+    consideration until the next iteration.
+
+    4) Repeat steps (1-3) until no more mergers occur.
+    """
+
+    anomalous_size = 800
+
+    # create a lookup table of OphysROIs
+    roi_lookup = {}
+    for roi in raw_roi_list:
+        roi_lookup[roi.roi_id] = roi
+
+    t0 = time.time()
+    logger.info('starting merger')
+
+    keep_going = True
+    incoming_rois = list(roi_lookup.keys())
+
+    have_been_merged = set()  # keep track of ROIs that were merged
+    valid_roi_id = set(roi_lookup.keys())
+
+    # pseudo-randomly order ROIs when calculating
+    # expensive statistics to prevent the most expensive
+    # ROIs from all landing on the same process
+    shuffler = np.random.RandomState(11723412)
+
+    merger_candidates = find_merger_candidates(list(roi_lookup.values()),
+                                               np.sqrt(2.0),
+                                               rois_to_ignore=None,
+                                               n_processors=n_processors)
+
+    # construct a dict mapping an ROI ID to the list of its
+    # neighboring ROI IDs (prevents us from having to call
+    # find_merger_candidates on every iteration)
+    neighbor_lookup = {}
+    for pair in merger_candidates:
+        if pair[0] not in neighbor_lookup:
+            neighbor_lookup[pair[0]] = set()
+        if pair[1] not in neighbor_lookup:
+            neighbor_lookup[pair[1]] = set()
+        neighbor_lookup[pair[1]].add(pair[0])
+        neighbor_lookup[pair[0]].add(pair[1])
+
+    logger.info('found global merger candidates in '
+                f'{time.time()-t0:.2f} seconds')
+
+    # global lookup tables for per-ROI sub-videos and mappings
+    # between potential merger pairs and merger metrics
+    sub_video_lookup = {}
+    merger_to_metric = {}
+    timeseries_lookup = {}
+
+    # ROIs that are too large to be considered valid
+    anomalous_rois = {}
+
+    logger.info(f'initially {len(roi_lookup)} ROIs')
+    while keep_going:
```

Comment:
I echo some of Adam's comments below. This is the main iterative loop, we should be able to read it and understand what's happening. As it stands there is too much exposed functionality right here to make that possible. Parting out that functionality to well-named functions will help.

Reply:
I've tried to split out each step in that loop into its own function. I can further condense things, if requested.